dan_layer/consensus/src/hotstuff/on_receive_proposal.rs

```diff
@@ -229,6 +229,57 @@ where TConsensusSpec: ConsensusSpec
         Ok(())
     }
 
+    fn on_receive_foreign_block(
+        &self,
+        tx: &mut <TConsensusSpec::StateStore as StateStore>::WriteTransaction<'_>,
+        block: &Block,
+        foreign_committee_shard: &CommitteeShard,
+    ) -> Result<(), HotStuffError> {
+        // Save the QCs if it doesnt exist already, we'll reference the QC in subsequent blocks
+        block.justify().save(tx)?;
+
+        // TODO(perf): n queries
+        for cmd in block.commands() {
+            let Some(t) = cmd.local_prepared() else {
+                continue;
+            };
+            let Some(mut tx_rec) = self.transaction_pool.get(tx, &t.id).optional()? else {
+                continue;
+            };
+
+            if tx_rec.stage().is_complete() {
+                warn!(
+                    target: LOG_TARGET,
+                    "⚠️ Foreign proposal received after transaction {} is COMPLETE. Ignoring.",
+                    tx_rec.transaction.id);
+                continue;
+            }
+
+            tx_rec.update_evidence(tx, foreign_committee_shard, *block.justify().id())?;
+            let change_to_abort = cmd.decision().is_abort() && tx_rec.original_decision().is_commit();
+            if change_to_abort {
+                info!(
+                    target: LOG_TARGET,
+                    "⚠️ Foreign shard ABORT {}. Update decision to ABORT",
+                    tx_rec.transaction.id
+                );
+                tx_rec.update_decision(tx, Decision::Abort)?;
+            }
+
+            // If we've received we know that all locals have prepared and have all the evidence from all shards, we can
+            // transition to All/SomePrepared
+            if tx_rec.stage().is_local_prepared() && tx_rec.transaction.evidence.all_shards_complete() {
```

Comment:
Only transistion when our local leader proposes, so that all local nodes have the same data